### PR TITLE
Fix layout mapping

### PR DESCRIPTION
**Pull request by OliverO2**
* Type: **Bug Fix**

* Impact: **Normal**

* Reference to related issue (URL): none

* How was this pull request tested? Ran `rear mkrescue` and `rear recover` on Ubuntu 18.04.2 LTS server before and after patching. Verified fix works as intended.

* Brief description of the changes in this pull request:

This PR corrects a mapping error when TCG Opal 2 self-encrypting disks were present but had to be excluded due to a non-existent disk during recovery. This exclusion was not handled properly.

(Edit: dropped second change)

### CODE REVIEW

File: usr/share/rear/lib/layout-functions.sh [hunk 1]
```diff
@@ -982,7 +982,7 @@ function apply_layout_mappings() {
     while read original replacement junk ; do
         # Skip lines that have wrong syntax:
         test "$original" -a "$replacement" || continue
-        if grep -q "$replacement" "$file_to_migrate" ; then
+        if grep -q "'^[^#].*$replacement" "$file_to_migrate" ; then
```
Comment (jsmeix) [inline usr/share/rear/lib/layout-functions.sh]: Please add a comment here that explains the reason behind that grep regexp
i.e. commented out lines that contain $replacement can be ignored and the reason
why commented out lines that contain $replacement can be ignored (see below).

The new regexp makes it no longer match at all for me
```
# replacement="_REAR123_"

# echo "# _REAR123_ foo bar" | grep "'^[^#].*$replacement" && echo y

# echo "_REAR123_ foo bar" | grep "'^[^#].*$replacement" && echo y
```
in contrast to the old regexp
```
# echo "_REAR123_ foo bar" | grep "$replacement" && echo y
_REAR123_ foo bar
y
```
so that I suggest simpler code instead of a too complicated all-in-one regexp
```
# echo "# _REAR123_ foo bar" | grep -v '^#' | grep "$replacement" && echo y

# echo "_REAR123_ foo bar" | grep -v '^#' | grep "$replacement" && echo y
_REAR123_ foo bar
y
```
i.e. code like
```
if grep -v '^#' "$file_to_migrate" | grep "$replacement" ; then
```

But I wonder why commented out lines can be ignored here at all
because in the apply_layout_mappings function the `sed` commands
in "Step1" and "Step 2" should not care about commented out lines
so that replacement hapens also in commented out lines
and then during "Step 3" no `_REAR*_` string should be left
in any line regardless whether it is commented out or not.

Simply put:
I think the `LogPrintError "Failed to apply layout mappings to ...`
that you see is correct and you would have to fix its root cause.

File: usr/share/rear/layout/prepare/default/300_map_disks.sh [hunk 2]
```diff
@@ -309,17 +309,23 @@ while true ; do
...
         is_mapping_source "$device" && continue
-        LogUserOutput "Disk $device and all dependant devices will not be recreated"
-        mark_as_done "$device"
-        mark_tree_as_done "$device"
-    done < <( grep -E "^disk |^multipath " "$LAYOUT_FILE" )
+        if [[ "$keyword" == "opaldisk" ]]; then
+            Log "TCG Opal 2-compliant self-encrypting disk $device will not be recreated"
+            # Note: dependent devices might still be recreated if $device is mapped as a (non-encrypting) disk
+            mark_as_done "opaldisk:$device"
+        else
```
Comment (jsmeix) [inline usr/share/rear/layout/prepare/default/300_map_disks.sh]: Only a personal coding style side note:
In loops I prefer to `continue` with the next loop iteration
instead of `if ... then ... else ... fi` clauses like
```
while ... ; do
    if SPECIAL_CASE. ; then
        ...
        continue
    fi
    ...
done
```
cf. the basic idea behind "Return early, return often"
in https://github.com/rear/rear/wiki/Coding-Style
but that is a matter of personal coding style taste.

**Comment by jsmeix** [comment]
@OliverO2 
I assume this issue hapens only when you are in MIGRATION_MODE
because the apply_layout_mappings function is only called there.

I would like to better understand what actually goes on on your
particular system because 'an unmapped disk was present' looks like
things are wrong because - as far as I can imagine offhandedly - there
must not be any unmapped disk left, see my comments in the
apply_layout_mappings function that show an example where
a `_REAR*_` string is left `disk _REAR3_` which is correctly
detected and reported by "Step 3".

File: usr/share/rear/lib/layout-functions.sh (hunk 1 above)
Comment (OliverO2) [reply usr/share/rear/lib/layout-functions.sh]: 1. There is a bug in my code: Somehow a single quote slipped in, which shouldn't have been there.
1.  Your test did not match because you tried a replacement at the beginning of a line. This is not supposed to happen in `disklayout.conf`, right?

If you correct my mistake and adjust the test it works:
```
replacement="_REAR123_"
echo "#disk _REAR123_ foo bar" | grep "^[^#].*$replacement" && echo y
echo "disk _REAR123_ foo bar" | grep "^[^#].*$replacement" && echo y
```

I have just tested a bit more and it seems like the case is not as obvious as I had initially thought it to be. I'll do some more testing and come up with results.

Comment (jsmeix) [reply usr/share/rear/lib/layout-functions.sh]: The apply_layout_mappings function can be called with any file,
not only with $LAYOUT_FILE so a `_REAR*_` string can be anywhere.

Currently we call the apply_layout_mappings function
in the following ways, note in particular the last one:

layout/prepare/default/320_apply_mappings.sh
```
apply_layout_mappings "$LAYOUT_FILE" ...
```

finalize/GNU/Linux/260_rename_diskbyid.sh
```
OLD_ID_FILE="${VAR_DIR}/recovery/diskbyid_mappings"
...
apply_layout_mappings "$OLD_ID_FILE" ...
```

finalize/GNU/Linux/250_migrate_disk_devices_layout.sh
```
for restored_file in [b]oot/{grub.conf,menu.lst,device.map} [e]tc/grub.* [b]oot/grub/{grub.conf,menu.lst,device.map} \
                     [b]oot/grub2/{grub.conf,grub.cfg,menu.lst,device.map} \
                     [e]tc/sysconfig/grub [e]tc/sysconfig/bootloader \
                     [e]tc/lilo.conf \
                     [e]tc/yaboot.conf \
                     [e]tc/mtab [e]tc/fstab \
                     [e]tc/mtools.conf \
                     [e]tc/smartd.conf [e]tc/sysconfig/smartmontools \
                     [e]tc/sysconfig/rawdevices \
                     [e]tc/security/pam_mount.conf.xml [b]oot/efi/*/*/grub.cfg
do
    ...
    ... apply_layout_mappings "$restored_file" ...
    ...
done
```

Comment (OliverO2) [reply usr/share/rear/lib/layout-functions.sh]: The original problem had appeared on bare metal systems equipped with TCG Opal 2 self-encrypting disks, so I did another round of tests on one of those. Migration mode was used as the target systems had a smaller number of disks than the original systems. With the mapping correction in place (commit fbfb9d6), the original problem (uncommented lines in disk layout containing `_REAR*_` placeholders) did not reappear.

So I have documented in my reversion commit:

The fix is unnecessary as the problem seemed to be caused
by a bug concerning the exclusion of of TCG Opal 2 self-encrypting
disks (now fixed by commit fbfb9d6).

Sorry for the annoyance.

File: usr/share/rear/layout/prepare/default/300_map_disks.sh (hunk 2 above)
Comment (OliverO2) [reply usr/share/rear/layout/prepare/default/300_map_disks.sh]: I'm OK with early returns to conditionally skip entire rear scripts. I'm also fine using goto-style for exception handling in languages without support for try/catch constructs.

Otherwise, I'd avoid using goto-like statements. Short-circuiting loops might easily be overlooked when adding code at the bottom of the loop later on.

Comment (jsmeix) [reply usr/share/rear/layout/prepare/default/300_map_disks.sh]: As I said: It is a matter of personal coding style taste.
I think @schlomo agrees with your pint of view that
```
short-circuiting loops might easily be overlooked
when adding code at the bottom of the loop later on
```
while my point of view contradicts because I think
```
whoever adds some code at the bottom of
"a closed piece of code" (e.g. a loop, an if clause, a function, ...)
cannot expect that his addon works as intended
unless he had examined what all the code above
his addon actually does
```
and therefore I prefer readability over all
because I must read and understand all the code
before I can do any changes.

**Comment by jsmeix** [comment]
@gdha 
could you also have a look here and review it?

Do you think we can add it to ReaR 2.5?

File: usr/share/rear/lib/layout-functions.sh (hunk 1 above)
Comment (jsmeix) [reply usr/share/rear/lib/layout-functions.sh]: No need to be sorry.
I appreciate all your various usage and testing of ReaR.

**Comment by gdha** [comment]
@jsmeix I'm fine for release 2.5,
@OliverO2 Thank you for improving rear code base.

**Comment by jsmeix** [comment]
Merged and listed in the ReaR 2.5 release notes via
https://github.com/rear/rear.github.com/commit/a78033bf1b97e0b3f638cf32c5f11ce5e049c085
